kore/src/Kore/Attribute/Pattern/FreeVariables.hs

```diff
@@ -83,21 +84,28 @@ mapFreeVariables
     :: Ord variable2
     => (variable1 -> variable2)
     -> FreeVariables variable1 -> FreeVariables variable2
-mapFreeVariables mapping (FreeVariables freeVariables) =
-    FreeVariables (Set.map (fmap mapping) freeVariables)
+mapFreeVariables mapping (FreeVariables freeVars) =
+    FreeVariables (Set.map (fmap mapping) freeVars)
 {-# INLINE mapFreeVariables #-}
 
 traverseFreeVariables
     :: (Applicative f, Ord variable2)
     => (variable1 -> f variable2)
     -> FreeVariables variable1 -> f (FreeVariables variable2)
-traverseFreeVariables traversing (FreeVariables freeVariables) =
+traverseFreeVariables traversing (FreeVariables freeVars) =
     FreeVariables . Set.fromList
-    <$> Traversable.traverse (traverse traversing) (Set.toList freeVariables)
+    <$> Traversable.traverse (traverse traversing) (Set.toList freeVars)
 {-# INLINE traverseFreeVariables #-}
 
 {- | Extracts the list of free element variables
 -}
 getFreeElementVariables :: FreeVariables variable -> [ElementVariable variable]
 getFreeElementVariables =
     mapMaybe extractElementVariable . Set.toList . getFreeVariables
+
+-- | Class for extracting the free variables of a pattern, term, rule, ...
+class HasFreeVariables pat variable where
```

Comment:
This should aid type inference greatly:

```suggestion
class HasFreeVariables thing where
    type FreeVariableType thing
    freeVariables :: thing -> FreeVariables (FreeVariableType thing)
```

Reply:
I tried to do that, but it does not seem to play nice with the instance below (`HasFreeVariables () variable`) as in that case the variable type does not depend on the pattern type. That is why I couldn't use functional dependencies in the first place.